crates/xline/src/conflict/tests.rs

```diff
@@ -181,14 +184,229 @@ fn exclusive_ucp_operations_are_ok() {
+    lease_collection.grant(1, 60, true);
+    lease_collection
+        .attach(1, "foo".as_bytes().to_vec())
+        .unwrap();
+
+    let kv_put = gen.gen_put("foo");
+    let kv_delete = gen.gen_delete_range("foo", "foz");
+    let lease_revoke = gen.gen_lease_revoke(1);
+
+    // put conflicts with lease revoke
+    assert!(!ucp.insert(kv_put.clone()));
+    assert!(ucp.insert(lease_revoke.clone()));
+    ucp.remove(&kv_put);
+    ucp.remove(&lease_revoke.clone());
+    assert!(!ucp.insert(lease_revoke.clone()));
+    ucp.remove(&lease_revoke.clone());
+
+    // delete range conflicts with lease revoke
+    assert!(!ucp.insert(kv_delete.clone()));
+    assert!(ucp.insert(lease_revoke.clone()));
+    ucp.remove(&kv_delete);
+    ucp.remove(&lease_revoke.clone());
+    assert!(!ucp.insert(lease_revoke.clone()));
+}
+
+#[test]
+fn ucp_revoke_then_kv_conflict_ok() {
+    let lease_collection = Arc::new(LeaseCollection::new(60));
+    let mut ucp = LeaseUncomPool::new(Arc::clone(&lease_collection));
+
+    let mut gen = EntryGenerator::default();
+    // suppose initially we have a key "foo" associated with lease id 1
+    lease_collection.grant(1, 60, true);
+    lease_collection
+        .attach(1, "foo".as_bytes().to_vec())
+        .unwrap();
+
+    let kv_put = gen.gen_put("foo");
+    let kv_delete = gen.gen_delete_range("foo", "foz");
+    let lease_revoke = gen.gen_lease_revoke(1);
+
+    // lease revoke conflicts with put
+    assert!(!ucp.insert(lease_revoke.clone()));
+    assert!(ucp.insert(kv_put.clone()));
+    ucp.remove(&lease_revoke.clone());
+    ucp.remove(&kv_put.clone());
+    assert!(!ucp.insert(kv_put.clone()));
+    ucp.remove(&kv_put.clone());
+
+    // lease revoke conflicts with delete range
+    assert!(!ucp.insert(lease_revoke.clone()));
+    assert!(ucp.insert(kv_delete.clone()));
+    ucp.remove(&lease_revoke.clone());
+    ucp.remove(&kv_delete.clone());
+    assert!(!ucp.insert(kv_delete.clone()));
+}
+
+/*
+The following test cases verify that insert and remove operations in a
+conflict pool are independent of any external state.
+
+Specifically, we need to query `LeaseCollection` for kv and lease requests
+and the `LeaseCollection` might be mutated sometime between a insert and
+a remove, potentially leading to an inconsist state in our conflict pool.
+*/
+
+#[test]
+fn kv_sp_mutation_no_side_effect() {
+    let lease_collection = Arc::new(LeaseCollection::new(60));
+    let mut sp = KvSpecPool::new(Arc::clone(&lease_collection));
+    let mut gen = EntryGenerator::default();
+
+    lease_collection.grant(1, 60, true);
+    lease_collection
+        .attach(1, "foo".as_bytes().to_vec())
+        .unwrap();
+    let kv_put = gen.gen_put("foo");
+    let lease_revoke = gen.gen_lease_revoke(1);
+
+    sp.insert_if_not_conflict(lease_revoke.clone());
+    assert!(sp.insert_if_not_conflict(kv_put.clone()).is_some());
+    lease_collection.detach(1, "foo".as_bytes()).unwrap();
+    sp.remove(&lease_revoke);
```

Comment:
What does this assertion prove? If you remove the lease_revoke from the sp, of course you won't get conflict while inserting the kv_put request.

Reply:
It aims to ensure that the `lease_revoke` command is correctly removed.